### ISSUE TITLE
Incorrect representation of empty has_one/has_many relationships

### ISSUE THREAD

**Issue by dmarkow** [AgilionApps/ja_serializer#26]
With a `has_one :author` relationship, if there is no linked author, you get this:

```json
"relationships": {
  "author": {
    "data": {
      "type": "author",
      "id": ""
    }
  }
}
```

However, the JSON API 1.0 spec says that resource linkage for an empty to-one relationship **must** be `null`, so I would expect:

```json
"relationships": {
  "author": {
    "data": null
  }
}
```

A `has_many :authors` scenario is slightly different. Right now, an empty `has_many` relationship gives you:

```json
"relationships": {
  "authors": {}
}
```

whereas the JSON API spec requires an empty array for empty to-many relationships:

```json
"relationships": {
  "authors": {
    "data": []
  }
}
```

If I have time this week I can start working on a PR, but wanted to raise it as an issue in the mean time. Thanks!

**Comment by dmarkow** [comment]
For has_many relationships, the cause was easily found. The `Utils` module's `put_if_present` function ignores both `nil` and `[]` and just returns the untouched `dict` object. In the relationship formatter, since calling `put_if_present` on  `:links` and `:data` does nothing to the dict, an empty dict is returned. Still not sure what the answer should be at this point, though.

For the has_one issue, we might be able to add a function clause in `JaSerializer.Builder.ResourceIdentifier` for `build` to handle the case where `id` is `nil`:

```elixir
def build(nil, _type), do: nil
```
However, this will then make it do the same thing as an empty has_many...

**Comment by dmarkow** [comment]
One possible approach might be to use `:null_value` and `:empty_list` atoms to help signal that we want to actually add explicit `null` and `[]` entries to the formatted output. We would add `put_if_present` clauses to actually add the null/empty values to `dict` if those atoms are used (while still ignoring actual `nil` and `[]` values):

```elixir
def put_if_present(dict, key, :null_value), do: Dict.put(dict, key, nil)
def put_if_present(dict, key, :empty_list), do: Dict.put(dict, key, [])
```

Then a few updates to `JaSerializer.Builder.ResourceIdentifier`. For has_many, add a case pattern for an empty array, and for has_one, a new function clause to handle a blank `id`:

```elixir
def build(%{serializer: serializer} = context, type, name) do
  serializer
  |> apply(name, [context.model, context.conn])
  |> case do
    [] -> :empty_list
    many when is_list(many) -> Enum.map(many, &build(&1, type))
    one -> build(one, type)
  end
end

def build(nil, _type), do: :null_value
```

I'm super new to Elixir and there may be a much better way to do this, so any feedback's appreciated.

**Comment by alanpeabody** [comment]
Amazingly I have not run into this yet. Thanks for reporting it!

So my take is that an "integration" test case is super important first step for this, from there we can experiment. (eg a test in json_api_spec that)

 I would like to try to stay away from using :null_value or :empty_list atoms because it feels a bit "un-elixirish". I think we should be able to use pattern matching and functions to address this.

**Comment by alanpeabody** [comment]
Fixed in #29.
